console_backend/src/ntripping.rs

```diff
@@ -0,0 +1,247 @@
+                let lon_min = (lonn - (lon_deg as f64)) * 60.0;
+
+                let lat_dir = if lat < 0.0 { 'S' } else { 'N' };
+                let lon_dir = if lon < 0.0 { 'W' } else { 'E' };
+
+                format!(
+                    "$GPGGA,{},{:02}{:010.7},{},{:03}{:010.7},{},4,12,1.3,{:.2},M,0.0,M,1.7,0078",
+                    time, lat_deg, lat_min, lat_dir, lon_deg, lon_min, lon_dir, height
+                )
+            }
+            Message::Cra {
+                request_counter,
+                area_id,
+                corrections_mask,
+                solution_id,
+            } => {
+                let mut s = String::from("$PSWTCRA,");
+                if let Some(request_counter) = request_counter {
+                    s.push_str(&format!("{request_counter},"));
+                }
+                s.push(',');
+                if let Some(area_id) = area_id {
+                    s.push_str(&format!("{area_id},"));
+                }
+                s.push(',');
+                if let Some(corrections_mask) = corrections_mask {
+                    s.push_str(&format!("{corrections_mask},"));
+                }
+                s.push(',');
+                if let Some(solution_id) = solution_id {
+                    s.push_str(&format!("{solution_id},"));
+                }
+                s
+            }
+        }
+    }
+}
+
+fn connect(opt: NtripOptions) -> anyhow::Result<()> {
+    let mut curl = Easy::new();
+    let mut headers = List::new();
+    headers.append("Transfer-Encoding:")?;
+    headers.append("Ntrip-Version: Ntrip/2.0")?;
+    headers.append(&format!("X-SwiftNav-Client-Id: {}", opt.client_id))?;
+    if opt.nmea_header {
+        if opt.area_id.is_some() {
+            headers.append(&format!("Ntrip-CRA: {}", build_cra(&opt)))?;
+        } else {
+            headers.append(&format!("Ntrip-GGA: {}", build_gga(&opt)))?;
+        }
+    }
+    curl.http_headers(headers)?;
+    curl.useragent("NTRIP ntrip-client/1.0")?;
+    curl.url(&opt.url)?;
+    curl.progress(true)?;
+    curl.put(true)?;
+    curl.custom_request("GET")?;
+    curl.http_version(HttpVersion::Any)?;
+
+    if let Some(username) = &opt.username {
+        curl.username(username)?;
+    }
+
+    if let Some(password) = &opt.password {
+        curl.password(password)?;
+    }
+    let (tx, rx) = channel::bounded::<Vec<u8>>(1);
+    let transfer = Rc::new(RefCell::new(curl.transfer()));
+
+    transfer.borrow_mut().progress_function({
+        let rx = &rx;
+        let transfer = Rc::clone(&transfer);
+        move |_dltot, _dlnow, _ultot, _ulnow| {
+            if !rx.is_empty() {
+                if let Err(e) = transfer.borrow().unpause_read() {
+                    eprintln!("unpause error: {e}");
+                    return false;
+                }
+            }
+            true
+        }
+    })?;
+
+    transfer.borrow_mut().write_function(|data| {
+        if let Err(e) = std::io::stdout().write_all(data) {
```

Comment:
The data sent to `io::stdout()` here will need to be sent as input to the connected device.  For Mallard, this will be RTCM, for other devices that only support SBP input we'll either need to disable the NTRIP feature entirely or figure out how to bundle an RTCM to SBP converter.